sympy/algebras/quaternion.py

```diff
@@ -78,8 +95,17 @@ class Quaternion(Expr):
 
     is_commutative = False
 
-    def __new__(cls, a=0, b=0, c=0, d=0, real_field=True):
-        a, b, c, d = map(sympify, (a, b, c, d))
+    def __new__(cls, a=0, b=0, c=0, d=0, real_field=True, norm=None):
+        a, b, c, d, norm = map(sympify, (a, b, c, d, norm))
+
+        if (norm is not None):
+            # if all elements are numerical
+            if (all(i.is_Number is True for i in [a, b, c, d]) and
+                    norm**2 != a**2 + b**2 + c**2 + d**2):
+                raise ValueError("incompatible value for norm")
+
+            if norm.is_zero:
```

Comment:
I think I would raise an error -- the norm can't be 0, can it? The user shouldn't be using 0 so it seems like they should just get an error right away.

Reply:
In theory, for `Quaternion(a, b, c, d)`, its norm is 0 if `a == b == c == d == 0`... Which is probably not a very useful quaternion.

Which one of these two options do you prefer:

1. Raise a `ValueError`, like you said
2. Give a warning a return a zero quaternion